### PR TITLE
Forms v2

### PR DESCRIPTION
**Pull request by Ghostvv**
**Proposed changes**:
- added new version of forms, that allows chitchat like interactions

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog

### CODE REVIEW

File: examples/formbot/data/stories.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+## happy path
+* request_restaurant
+    - restaurant_form
+    - form_activated{"form_name": "restaurant_form"}
```
Comment (tmbo) [inline examples/formbot/data/stories.md]: random thought, would 
```markdown
- form{"form_name": "restaurant_form"}
<!-- ... -->
- form{"form_name": null}
```
work as well?

File: examples/formbot/train.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+if __name__ == '__main__':
+    utils.configure_colored_logging(loglevel="INFO")
+
+    training_data_file = 'data/stories.md'
+    model_path = 'models/dialogue'
+
+    agent = Agent("domain.yml",
+                  policies=[AugmentedMemoizationPolicy(),
+                            FormPolicy(),
```
Comment (tmbo) [inline examples/formbot/train.py]: we can add the formpolicy by default though, right? (so you do not need to customize this if you want to use forms)

File: examples/moodbot/myslots.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+
+from rasa_core_sdk.slots import Slot
+
+class MyFreeTextSlot(Slot):
+
+    def name(self):
+        return "my_free_text_slot"
+
+    def validate(self, parse_data):
+        new_parse_data = parse_data.copy()
```
Comment (tmbo) [inline examples/moodbot/myslots.py]: I thought we don't do custom slots?

File: rasa_core_sdk/forms.py [hunk 4]
```diff
@@ -95,85 +39,64 @@ def name(self):
...
+        for e in entities:
+            if e.get("entity") == tracker.slots[REQUESTED_SLOT]:
+                events.append(SlotSet(e['entity'], e['value']))
+        if events:
+            return events
         else:
-            fields = [f
-                      for f in required
-                      if f.slot_name == requested_slot]
+            raise InputValidationError("validation not implemented")
```
Comment (tmbo) [inline rasa_core_sdk/forms.py]: I am not sure how this error message relates to the fact that there are no events?

File: examples/formbot/data/stories.md (hunk 1 above)
Comment (Ghostvv) [reply examples/formbot/data/stories.md]: I think so

File: examples/formbot/train.py (hunk 2 above)
Comment (Ghostvv) [reply examples/formbot/train.py]: what do you mean? add to `train.py`, or always add it?

File: examples/moodbot/myslots.py (hunk 3 above)
Comment (Ghostvv) [reply examples/moodbot/myslots.py]: yes, the whole thing should be deleted

File: examples/formbot/data/stories.md (hunk 1 above)
Comment (akelad) [reply examples/formbot/data/stories.md]: yeah i can change it to be that i think, if we'd rather that than `form:`

File: examples/formbot/train.py (hunk 2 above)
Comment (akelad) [reply examples/formbot/train.py]: yeah i'd say so, no reason not to always include it in train.py

File: rasa_core_sdk/forms.py (hunk 4 above)
Comment (akelad) [reply rasa_core_sdk/forms.py]: hahaha yes we already said we'd rename this. But wait did we say we're changing the error? I thought only the event?

Comment (Ghostvv) [reply rasa_core_sdk/forms.py]: I thought we create ActionExecutionError, that can be returned by any action

Comment (akelad) [reply rasa_core_sdk/forms.py]: i thought we were just changing the `ValidationFailed` event to `ActionExecutionFailed`?

Comment (Ghostvv) [reply rasa_core_sdk/forms.py]: but if we make event general, why not rename the error as well to sync it with event?

File: examples/formbot/data/stories.md (hunk 1 above)
Comment (tmbo) [reply examples/formbot/data/stories.md]: no this is not about the form prefix for ignored events but the events to activate / deactivate a form

Comment (Ghostvv) [reply examples/formbot/data/stories.md]: good idea, so we'd have only one event instead of two

Comment (Ghostvv) [reply examples/formbot/data/stories.md]: done

File: rasa_core_sdk/__init__.py [hunk 5]
```diff
@@ -147,13 +153,23 @@ def name(self):
...
         """Execute the side effects of this action."""
 
-        raise NotImplementedError
+        raise NotImplementedError("An action must implement its run method")
 
     def __str__(self):
         return "Action('{}')".format(self.name())
+
+
+class ActionExecutionError(Exception):
```
Comment (tmbo) [inline rasa_core_sdk/__init__.py]: rename as commented in core PR

File: rasa_core_sdk/endpoint.py [hunk 6]
```diff
@@ -69,7 +69,14 @@ def health():
     def webhook():
         """Webhook to retrieve action calls."""
         action_call = request.json
-        response = executor.run(action_call)
+        try:
+            response = executor.run(action_call)
+        except ActionExecutionError as e:
+            logger.error(str(e))
```
Comment (tmbo) [inline rasa_core_sdk/endpoint.py]: `from builtins import str`

File: rasa_core_sdk/endpoint.py [hunk 7]
```diff
@@ -69,7 +69,14 @@ def health():
...
+            result = {"error": str(e), "action_name": e.action_name}
```
Comment (tmbo) [inline rasa_core_sdk/endpoint.py]: `"name"` instead of `"action_name"`? not sure which one is more consistent

File: rasa_core_sdk/events.py [hunk 8]
```diff
@@ -139,3 +139,12 @@ def AgentUttered(text=None, data=None, timestamp=None):
         "data": data,
         "timestamp": timestamp,
     }
+
+
```
Comment (tmbo) [inline rasa_core_sdk/events.py]: not that it makes a lot of sense, but also needs the actionexecutionfailed event

Comment (Ghostvv) [reply rasa_core_sdk/events.py]: done

File: rasa_core_sdk/forms.py [hunk 9]
```diff
@@ -5,176 +5,197 @@
...
+    def submit(self, dispatcher, tracker, domain):
+        # type: (CollectingDispatcher, Tracker, Dict[Text, Any]) -> List[Dict]
+        """Define what the form has to do
+            after all required slots are filled"""
 
-    def __init__(self, slot_name, affirm_intent, deny_intent):
-        self.slot_name = slot_name
-        self.affirm_intent = affirm_intent
-        self.deny_intent = deny_intent
+        raise NotImplementedError("A form must implement a submit method")
```
Comment (Ghostvv) [inline rasa_core_sdk/forms.py]: @amn41 why `submit` raises `NotImplementedError`, why don't we use `pass` as default?

File: rasa_core_sdk/endpoint.py (hunk 6 above)
Comment (Ghostvv) [reply rasa_core_sdk/endpoint.py]: done

File: rasa_core_sdk/forms.py (hunk 9 above)
Comment (amn41) [reply rasa_core_sdk/forms.py]: I think in almost all cases you want to do something with the information once you collect it, this reminds the user that this `submit` will be called. If they really don't need it they can override it with a `pass`

**Comment by Ghostvv** [comment]
@MetcalfeTom could you please give examples of additional logs you need

File: rasa_core_sdk/endpoint.py (hunk 7 above)
Comment (Ghostvv) [reply rasa_core_sdk/endpoint.py]: let's keep `action_name` because it is `action_name` in the error

**Comment by Ghostvv** [comment]
yes, it is impossible, because requested slot is set by form action after it is executed

**Comment by Ghostvv** [comment]
log in `request_next_slot()` will be in action server

**Comment by Ghostvv** [comment]
@MetcalfeTom I added logger statements to `forms.py`

File: rasa_core_sdk/__init__.py [hunk 10]
```diff
@@ -170,7 +177,21 @@ def run(self, dispatcher, tracker, domain):
...
+
+
+class ActionExecutionRejection(Exception):
+    """Raising this exception will allow other policies
+        to predict another action"""
+
+    def __init__(self, action_name, message=None):
+        self.action_name = action_name
+        self.message = (message or
+                        "Custom action '{}' rejected to run"
```
Comment (akelad) [inline rasa_core_sdk/__init__.py]:
```suggestion
                        "Custom action '{}' rejected execution of"
```

File: rasa_core_sdk/forms.py [hunk 11]
```diff
@@ -5,176 +5,307 @@
...
-        slot_events = []
-        extracted_entities = {requested_entity}
+    # noinspection PyMethodMayBeStatic
+    def slot_mappings(self):
+        # type: () -> Dict[Text: Union[Dict, List[Dict]]]
+        """A dictionary to map required slots to
+            - an extracted entity
+            - intent: value pairs
+            - a whole message
+            or a list of them, where a first match will be picked
```
Comment (akelad) [inline rasa_core_sdk/forms.py]:
```suggestion
            or a list of them, where the first match will be picked
```

File: rasa_core_sdk/forms.py [hunk 12]
```diff
@@ -5,176 +5,307 @@
...
+
+    # noinspection PyUnusedLocal
+    def extract_other_slots(self,
+                            dispatcher,  # type: CollectingDispatcher
+                            tracker,  # type: Tracker
+                            domain  # type: Dict[Text, Any]
+                            ):
+        # type: (...) -> Dict[Text: Any]
+        """Extract the values of the other slots
+            if they are set by corresponded entities from a user input
```
Comment (akelad) [inline rasa_core_sdk/forms.py]:
```suggestion
            if they are set by corresponding entities from the user input
```

File: rasa_core_sdk/forms.py [hunk 13]
```diff
@@ -5,176 +5,307 @@
...
+                                 "for extra slot '{}'"
+                                 "".format(value, slot))
+                    slot_values[slot] = value
+
+        return slot_values
+
+    def validate(self, dispatcher, tracker, domain):
+        # type: (CollectingDispatcher, Tracker, Dict[Text, Any]) -> List[Dict]
+        """Validate extracted value of requested slot
+            else reject to execute the form action
```
Comment (akelad) [inline rasa_core_sdk/forms.py]:
```suggestion
            else reject execution of the form action
```

File: rasa_core_sdk/forms.py [hunk 14]
```diff
@@ -5,176 +5,307 @@
...
+            logger.debug("There is no active form")
 
-        if requested_slot is None:
+        if tracker.active_form.get('name') == self.name():
             return []
         else:
-            fields = [f
-                      for f in required
-                      if f.slot_name == requested_slot]
+            logger.debug("Activate the form '{}'".format(self.name()))
```
Comment (akelad) [inline rasa_core_sdk/forms.py]:
```suggestion
            logger.debug("Activated the form '{}'".format(self.name()))
```

File: rasa_core_sdk/forms.py [hunk 15]
```diff
@@ -5,176 +5,307 @@
...
+        # type: (CollectingDispatcher, Tracker, Dict[Text, Any]) -> List[Dict]
+        """Return a list of events from `self.validate(...)`
+            if validation is required:
+            - the form is active
+            - the form is called after `action_listen`
+            - form validation was not cancelled
+        """
+        if (tracker.latest_action_name == 'action_listen' and
+                tracker.active_form.get('validate', True)):
+            logger.debug("Validate user input")
```
Comment (akelad) [inline rasa_core_sdk/forms.py]:
```suggestion
            logger.debug("Validating user input")
```

File: rasa_core_sdk/forms.py [hunk 16]
```diff
@@ -5,176 +5,307 @@
...
+            return self.validate(dispatcher, tracker, domain)
+        else:
+            logger.debug("Skip validation")
```
Comment (akelad) [inline rasa_core_sdk/forms.py]:
```suggestion
            logger.debug("Skipping validation")
```

File: rasa_core_sdk/forms.py [hunk 17]
```diff
@@ -5,176 +5,307 @@
...
+        return tracker.get_slot(slot_name) is None
 
-        events = (self.get_requested_slot(tracker) +
-                  self.get_other_slots(tracker))
+    @staticmethod
+    def _deactivate():
+        # type: () -> List[Dict]
+        """Return `Form` event with `None` as name to deactivate the form
+            and reset the requested slot"""
+        logger.debug("Deactivating the form")
```
Comment (akelad) [inline rasa_core_sdk/forms.py]: add the name to the logger message?

File: rasa_core_sdk/forms.py [hunk 18]
```diff
@@ -5,176 +5,307 @@
...
-        raise NotImplementedError
+        """Unique identifier of the form"""
 
-    RANDOMIZE = True
+        raise NotImplementedError("A form must implement a name")
 
     @staticmethod
-    def required_fields():
-        return []
+    def required_slots(tracker):
```
Comment (akelad) [inline rasa_core_sdk/forms.py]: can we add a comment to this that you can modify this function to request conditional slots as well? so like not just required ones, but if certain slots are filled you can ignore some other slots

File: rasa_core_sdk/forms.py (hunk 17 above)
Comment (Ghostvv) [reply rasa_core_sdk/forms.py]: 👍

File: rasa_core_sdk/forms.py (hunk 18 above)
Comment (Ghostvv) [reply rasa_core_sdk/forms.py]: added a phrase

File: rasa_core_sdk/forms.py [hunk 19]
```diff
@@ -5,176 +5,397 @@
...
+                    intent=None,  # type: Optional[Union[Text, List[Text]]]
+                    not_intent=None  # type: Optional[Union[Text, List[Text]]]
+                    ):
+        # type: (...) -> Dict[Text: Any]
+        """A dictionary for slot mapping to extract slot value from
+            - an extracted entity
+            - conditioned on
+                - intent if it is not None
+                - not_intent if it is not None,
+                    meaning user intent should be different
```
Comment (akelad) [inline rasa_core_sdk/forms.py]:
```suggestion
                    meaning user intent should not be this intent
```

File: rasa_core_sdk/forms.py [hunk 20]
```diff
@@ -5,176 +5,397 @@
...
+                    intent=None,  # type: Optional[Union[Text, List[Text]]]
+                    not_intent=None  # type: Optional[Union[Text, List[Text]]]
+                    ):
+        # type: (...) -> Dict[Text: Any]
+        """A dictionary for slot mapping to extract slot value from
+            - intent: value pair
+            - conditioned on
+                - intent if it is not None
+                - not_intent if it is not None,
+                    meaning user intent should be different
```
Comment (akelad) [inline rasa_core_sdk/forms.py]:
```suggestion
                    meaning user intent should not be this intent
```

File: rasa_core_sdk/forms.py [hunk 21]
```diff
@@ -5,176 +5,397 @@
...
+                  intent=None,  # type: Optional[Union[Text, List[Text]]]
+                  not_intent=None  # type: Optional[Union[Text, List[Text]]]
+                  ):
+        # type: (...) -> Dict[Text: Any]
+        """A dictionary for slot mapping to extract slot value from
+            - a whole message
+            - conditioned on
+                - intent if it is not None
+                - not_intent if it is not None,
+                    meaning user intent should be different
```
Comment (akelad) [inline rasa_core_sdk/forms.py]:
```suggestion
                    meaning user intent should not be this intent
```

File: rasa_core_sdk/forms.py [hunk 22]
```diff
@@ -5,176 +5,397 @@
...
+            if (not isinstance(requested_slot_mapping, dict) or
+                    requested_slot_mapping.get("type") is None):
+                raise TypeError("Provided incompatible slot mapping")
+
+        return requested_slot_mappings
 
-    def get_requested_slot(self, tracker):
-        # type: (Tracker) -> List[EventType]
+    @staticmethod
+    def intent_is_desired(requested_slot_mapping, tracker):
```
Comment (akelad) [inline rasa_core_sdk/forms.py]: types?

**Pull request merged**